src/lib.rs

```diff
@@ -1,101 +1,49 @@
-//! Simple LDAP -> Famedly Zitadel sync tool to match users between
-//! clients and our infrastructure.
-use std::path::Path;
+//! Sync tool to match users between clients and our infrastructure.
 
-use anyhow::{bail, Context, Result};
-use ldap_poller::{ldap::EntryStatus, ldap3::SearchEntry, Cache, Ldap};
-use tokio::sync::mpsc::Receiver;
-use tokio_stream::{wrappers::ReceiverStream, StreamExt};
+use anyhow::Result;
 
 mod config;
+mod sync_endpoint;
+mod sync_ldap;
 mod user;
 mod zitadel;
 
 pub use config::{AttributeMapping, Config, FeatureFlag};
+use sync_endpoint::EndpointSync;
+use sync_ldap::LdapSync;
 use zitadel::Zitadel;
 
-/// Run the sync
-pub async fn sync_ldap_users_to_zitadel(config: Config) -> Result<()> {
+/// Trait to define syncing operations for different sources
+trait SyncSource {
+	/// Create a new sync source
+	fn new(config: &Config) -> Result<Self>
+	where
+		Self: Sized;
+
+	/// Perform a sync operation
+	async fn sync(&self, zitadel_client: &Zitadel) -> Result<()>;
+}
```

Comment:
I don't think this version of `SyncSource` is what we need. Doing this way would put too much behind implementation. We have a lot of stuff that could be source-independent: the all of zitadel interaction and caching

If it's `*Source` I'd say it should be treated as a resource abstraction, having get and set (well, we don't need the latter in our case) method. Looking at this design I'd rather call it `SyncMethod` or something like that, but not source

If it is called "source" it shouldn't have any capabillity to interact with zitadel, as zitadel is sink, `SyncSource` should just be a source abstraction